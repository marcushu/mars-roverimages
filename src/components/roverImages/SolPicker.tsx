import { FunctionComponent, useState } from "react";
import NextBackBtn from "./NextBackBtn";
import styles from './solPicker.module.css';

interface SolPickerProps {
  currentSol: number
  setSol: React.Dispatch<React.SetStateAction<number>>
}

const SolPicker: FunctionComponent<SolPickerProps> = ({ currentSol, setSol }) => {
  const [solValue, setSolValue] = useState(1);

  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
    const num = parseInt(evt.target.value);
    //if solValue !== currentSol
    //  hide next/back buttons
    //  display GoTo button
    //willl require a handler (prop) to set new page
    setSolValue(num);
  }

  const goToSol = () => {
    alert('go to ' + solValue.toString());
  }

  const nextSol = () => {
    setSol(current => current + 1);
  }

const prevSol = () => {
  if(currentSol > 2)
    setSol(current => current - 1);
}

  return (
    <div className={styles.solpicker}>
      <NextBackBtn goFunction={prevSol}>
        &#10094;
      </NextBackBtn>
      sol <input type="number" value={solValue} onChange={e => handleChange(e)} /> of 2341
      <NextBackBtn goFunction={solValue === 3 ? goToSol : nextSol}>
        &#10095;
      </NextBackBtn>
    </div>
  );
}

export default SolPicker;