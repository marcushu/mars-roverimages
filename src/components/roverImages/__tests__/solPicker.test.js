import { fireEvent, render, screen } from "@testing-library/react";
import SolPicker from "../SolPicker";

const mockFunc = jest.fn();


it('should display correct sol and total sols', () => {
    render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);

    // total sols
    const solInput = screen.getByText(/solof 1234/i);
    expect(solInput).toBeInTheDocument();
    //current sol
    expect(screen.getByDisplayValue(3)).toBeInTheDocument();
});


it('should only call the setsol function if the input is not empty', () => {
    render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);

    const solInput = screen.getByRole('spinbutton');

    // empty input...
    fireEvent.change(solInput, {target: {value: ''}});
    const rightBtn = screen.getByRole('button', {  name: /❯/i});
    fireEvent.click(rightBtn);
    expect(mockFunc.mock.calls).toHaveLength(0)

    // with input
    fireEvent.change(solInput, {target: {value: '10'}});
    fireEvent.click(rightBtn);
    expect(mockFunc.mock.calls).toHaveLength(1);
});


it('should use next-button to call the function with the correct input', () => {
    render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);
    
    // click next, no input change, should go to next page
    const rightBtn = screen.getByRole('button', {  name: /❯/i});
    fireEvent.click(rightBtn);

    expect(mockFunc.mock.calls[0][0]).toBe(4);
});


it('should use previous-button to call the function with the correct input', () => {
    render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);

    // click previous, no input change, should go to previos page
    const leftBtn = screen.getByRole('button', {  name: /❮/i});
    fireEvent.click(leftBtn);

    expect(mockFunc.mock.calls[0][0]).toBe(2);
});


it('should call the function with the correct jump to value', () => {
    render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);

    const solInput = screen.getByRole('spinbutton');
    fireEvent.change(solInput, {target: {value: '100'}});

    // next button
    const rightBtn = screen.getByRole('button', {  name: /❯/i});
    fireEvent.click(rightBtn);

    expect(mockFunc.mock.calls[0][0]).toBe(100);

    // prev button
    const leftBtn = screen.getByRole('button', {  name: /❮/i});
    fireEvent.click(leftBtn);

    expect(mockFunc.mock.calls[0][0]).toBe(100);
});


it('should match snapshot', () => {
    const { asFragment } = render(<SolPicker currentSol={3} totalSols={1234} setSol={mockFunc} />);

    expect(asFragment()).toMatchSnapshot();
});
