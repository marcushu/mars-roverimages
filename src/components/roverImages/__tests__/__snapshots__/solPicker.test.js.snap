// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="solpicker"
  >
    <button>
      ❮
    </button>
    sol
    <input
      type="number"
    />
    of 1234
    <button>
      ❯
    </button>
  </div>
</DocumentFragment>
`;
